### PR TITLE
[FLINK-3738] [tableAPI] Refactor TableEnvironments. Remove Translators and TranslationContext.

### PR DESCRIPTION
**Pull request by fhueske**
- [x] General
  - The pull request references the related JIRA issue
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message

- [X] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [X] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

- Removes Translators (JavaBatchTranslator, ScalaBatchTranslator, JavaStreamTranslator, ScalaStreamTranslator)
- Removes TranslationContext
- Creates TableEnvironments for Java/Scala DataSet/DataStream APIs
- Functionality of translators and translation context is moved to table environments
- Tables are bound to a TableEnvironment
- Simplifies table translation from and to DataSet / DataStream.
- Updates documentation and cleaned up ScalaDocs

### CODE REVIEW

**Comment by twalthr** [comment]
Thanks for the massive refactoring @fhueske!
I looked over the changes and couldn't spot anything noteworthy. The only things that I don't like is the additional `TableEnvironment` creattion even for simple table programs when using implicit conversions. Couldn't we just also provide a static TableEnvironment if we call the parameterless `toTable` that just calls `(Stream)ExecutionEnvironment.getExecutionEnvironment()`?

**Comment by fhueske** [comment]
I thought about that as well, but decided against an implicit conversion without TableEnvironment.

What might happen otherwise is that users try to join or union tables that have different implicitly created TableEnvironments or to register an implicitly converted table in a TableEnvironment. This does not work due to Calcite internals. I think it is better to make the Table/TableEnvironment relationship explicit and also easier to understand for users if an exception is thrown for the above reason.

File: docs/apis/batch/libs/table.md [hunk 1]
```diff
@@ -61,58 +59,73 @@ Note that the Table API is currently not part of the binary distribution. See li
 
 Table API
 ----------
-The Table API provides methods for running relational operations on Tables, both in Scala and Java.
-In the following sections you can find examples that show how to create Tables, how to define and execute relational queries on them,
-and how to retrieve the result of a query as a `DataSet`.
+The Table API provides methods to apply relational operations on DataSets, both in Scala and Java.
+
+The central concept of the Table API is a `Table` which is represents a table with relational schema (or relation). Tables can be created from a `DataSet`, converted into a `DataSet`, or registered in a table catalog using a `TableEnvironment`. A `Table` is always bound to a specific `TableEnvironment`. It is not possible to combine Tables of different TableEnvironments.
```
Comment (vasia) [inline docs/apis/batch/libs/table.md]: Remove "is".

File: docs/apis/batch/libs/table.md [hunk 2]
```diff
@@ -61,58 +59,73 @@ Note that the Table API is currently not part of the binary distribution. See li
...
-description of the expression syntax.
+Please refer to the Scaladoc (and Javadoc) for a full list of supported operations and a description of the expression syntax.
 
 {% top %}
 
 ### Java Table API
 
-When using Java, Tables can be converted to and from DataSet using `TableEnvironment`.
-This example is equivalent to the above Scala Example:
+When using Flink's Java DataSet API, DataSet are converted to Tables and Tables to DataSet using a `TableEnvironment`.
```
Comment (vasia) [inline docs/apis/batch/libs/table.md]: *DataSets

File: docs/apis/batch/libs/table.md [hunk 3]
```diff
@@ -391,9 +375,7 @@ val result = in.distinct();
 {% top %}
 
 ### Expression Syntax
-Some of operators in previous section expect an expression. These can either be specified using an embedded Scala DSL or
-a String expression. Please refer to the examples above to learn how expressions can be
-formulated.
+Some of operators in previous section expect one or more expressions. Expressions can be specified using an embedded Scala DSL or as Strings. Please refer to the examples above to learn how expressions can be specified.
```
Comment (vasia) [inline docs/apis/batch/libs/table.md]: Some of _the_ operators or Some operators
*sections

**Comment by vasia** [comment]
Just a couple of typos from my side.
I really like it that we got rid of the `TranslationContext` and I'm in favor of the explicit relationship between Tables and TableEnvironment. +1 to merge!

**Comment by fhueske** [comment]
Thanks for the review @vasia.
I fixed the typos.

**Comment by fhueske** [comment]
will merge this.

**Pull request closed without merging**